API changes to extend custom presentation

danigm opened the pull request:
## Summary

Extend the custom content renderer javascript API to be able to do advanced content query. This is the list of changes made to the API:

 * Add pagination parameters to `getContentByFilter` and `searchContent` queries and responses.
 * Filter all queries in the API by the channel so there's no nodes outside of the channel scope.
 * Add more options to `getContentByFilter`: onlyTopics, onlyContent, kinds, authors, tags and random
 * Add node thumbnail url to the node serialization
 * Add a new `getChannelMetadata` API method
 * Add a new  `getChannelFilterOptions` API method to get available authors, content kinds and tags in a channel
 * New random filter for content nodes that sorts the nodes randomly
 * New filter by tags for content nodes

----

## Testing checklist

- [x] Contributor has fully tested the PR manually. We are updating our custom presentation for the Endless Key using this new API and we're testing all the new methods and calls.

## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

danigm committed:
CustomContent: Add option to fetch only content

danigm committed:
CustomContent: Use new cursor pagination in API

https://phabricator.endlessm.com/T32573

danigm committed:
CustomContent: filter by channel

This patch adds the channel_id filter to the search and
getChannelByFilter queries to limit the results to the current channel.

danigm committed:
CustomContent: Remove unused under option from Search

The "under" option is defined in the Search API but is not implemented
in the real call.

danigm committed:
Add kinds option to kolibri.getContentByFilter

danigm committed:
Add authors option to kolibri.getContentByFilter

This new option will use the new author filter that filter the nodes by
author.

manuq committed:
CustomContent: Add option to fetch only topics

Add a onlyTopics option to the fetchContentCollection method. This is
a shortcut to pass `kind: topic` to the kolibri API.

danigm committed:
Add tags option to kolibri.getContentByFilter

This new option will use the new tags filter that filter the nodes by
tags.

manuq committed:
API: Add node thumbnail when processing nodes

Lookup the files and if there is a thumbnail, add it's storage_url to
the node item.

danigm committed:
Add new channel metadata API method

danigm committed:
Add new kolibri API method getChannelFilterOptions

This new method returns some channel aggregated data about the content
on that channel to be used later un getContentByFilter.

In this implementation there are three values:
 * availableAuthors
 * availableTags
 * availableKinds

rtibbles commented on kolibri/core/content/api.py:
```diff
@@ -455,6 +455,11 @@ def process_items(self, items, queryset, ancestor_lookup_method=None):
             if ancestor_lookup_method:
                 item["ancestors"] = ancestor_lookup_method(item)
             item["is_leaf"] = item.get("kind") != content_kinds.TOPIC
+            children = models.ContentNode.objects.filter(parent=item["id"])
```
This will cause a massive performance regression, btw - so we cannot merge this as is.

danigm replied:
I've just removed this commit for now. We want to know the number of children and topic children to be able to detect the last level in the tree (if it has no topic children) and also to show some information in topic nodes, the number of children.

But we can rethink all of this frontend functionality if we don't find an efficient way to add this information to nodes.

rtibbles commented:
See here for the latest updates to the contentnode backend re pagination and metadata: https://github.com/learningequality/kolibri/pull/8488

manuq commented:
> See here for the latest updates to the contentnode backend re pagination and metadata: #8488

OK, that one got merged so we should rebase then.

danigm commented:
I've detected that the random filter in this PR is not random anymore with the new pagination system. I'll take a look to this to try to solve.

danigm committed:
Add new kolibri API method getRandomNodes

danigm commented:
> I've detected that the random filter in this PR is not random anymore with the new pagination system. I'll take a look to this to try to solve.

After some testing, I can confirm that the random filter in combination with the pagination is not a solution anymore, so I've implemented the random query as a new API method: https://github.com/endlessm/kolibri/commit/5c94506f906638cea8a4fb3f0ed6d0533515cac6

rtibbles commented on kolibri/plugins/learn/assets/src/views/ChannelRenderer/CustomContentRenderer.vue:
```diff
@@ -114,19 +132,33 @@
+            authors: options.authors,
+            tags: options.tags,
             parent: options.parent === 'self' ? this.topic.id : options.parent,
+            channel_id: this.topic.channel_id,
```
The parent option should be used for filtering here - this will prevent custom renderers from displaying any content from outside their channels.

Alternatively, can add a `sameChannel` option to toggle this.

danigm replied:
I've added a new option `limitToChannel` to toggle this. With the parent we limit to just direct children, but with this option we can get more than one level at once, so this is more flexible.

rtibbles commented on kolibri/plugins/learn/assets/src/views/ChannelRenderer/CustomContentRenderer.vue:
```diff
@@ -114,19 +132,33 @@
+            max_results: options.maxResults ? options.maxResults : 50,
+            cursor: options.cursor,
+            kind: onlyTopics ? ContentNodeKinds.TOPIC : undefined,
+            kind_in: onlyContent ? allButTopicTypes : kinds,
```
I think this would be better done as an exclude filter on topic.

danigm replied:
I've discovered that this filter is already implemented in the `kind` filter, we just need to pass the value `content`. So I've updated the logic to use that filter instead of using the `kind_in` or create a new one.

rtibbles commented on kolibri/plugins/learn/assets/src/views/ChannelRenderer/CustomContentRenderer.vue:
```diff
@@ -114,19 +132,33 @@
             parent: options.parent === 'self' ? this.topic.id : options.parent,
+            channel_id: this.topic.channel_id,
+            max_results: options.maxResults ? options.maxResults : 50,
+            cursor: options.cursor,
```
I think rather than passing `cursor` here - it would be better to define a separate `fetchMore` pathway, and the `more` object that was returned can just be returned back. Otherwise, mixing the cursor with potential other changes to parameters could lead to unexpected results.

danigm replied:
Okay, I didn't thought about that problem. I've created a new method in the API, `getContentPage`, to request new pages with the cursor and just removed the `cursor` from the `getContentByFilter`.

rtibbles commented on kolibri/plugins/learn/assets/src/views/ChannelRenderer/CustomContentRenderer.vue:
```diff
@@ -245,6 +281,75 @@
         return this.hashi.mediator.sendMessage(newMsg);
       },
+      sendChannelMetadata(message) {
+        return ChannelResource.fetchModel({ id: this.topic.channel_id })
```
We can retrieve this from bootstrapped metadata now, see this composable: https://github.com/learningequality/kolibri/blob/release-v0.15.x/kolibri/plugins/learn/assets/src/composables/useChannels.js

This will avoid a request.

danigm replied:
I've just updated this API call to use the `useChannels` function instead of making a request to the backend.

rtibbles commented on kolibri/plugins/learn/assets/src/views/ChannelRenderer/CustomContentRenderer.vue:
```diff
@@ -245,6 +281,75 @@
+          getParams: {
+            parent: options.parent === 'self' ? this.topic.id : options.parent,
+            channel_id: this.topic.channel_id,
+            max_results: options.maxResults ? options.maxResults : 10,
```
Pagination forces an ordering, so I am guessing this is not working.

danigm replied:
This new API method doesn't use the cursor pagination, it doesn't have any pagination, the `max_results` is only used to limit the number of nodes, but it's not possible to get more with the same "random" order, each call should provide a new set of randoms.

https://github.com/learningequality/kolibri/pull/8476/commits/6225201694aa59aa129bd9f97f4acb06656dfcc7#diff-f2b97d70d258e7c248156962c7e63fafe868abced1c0644b8b25a0aa1d0addd2

rtibbles reviewed (changes requested):
Mostly seems mergeable, but some tweaks that I think would be needed.

danigm committed:
CustomContent: Add option to toggle the channel_id filter

The new option (limitToChannel) can be used to toggle the request
filtering by channel id. It's set by default to true.

danigm committed:
CustomContent: Use kind filter for onlyContent option

The kind filter implements this functionality directly if the value is
'content', so we can use that instead of using the kind_in filter with
all the kinds except the topic one.

danigm committed:
CustomContent: Add a new method to get next pages

Instead of using the cursor in the fetchContentCollection it's better to
do it in a different API call to avoid the possibility of changing the
fitlers for a request with a cursor.

danigm committed:
CustomContent: Do not call to the API for the channel metadata

This patch uses the `useChannels` function to retrieve the channel
metadata intead of doing a call to the backend.

rtibbles commented:
Some linting issues, can run:
`pre-commit run --all-files` and commit the results to fix.

danigm committed:
Fix lint errors

rtibbles reviewed (approved):
This all seems reasonable to me - doesn't seem like a huge maintenance burden by continuing to support all of these.

rtibbles commented:
Linting failure has been fixed on release branch, merging.
